README.md
# ibmAcousticR
Connect to Your 'IBM Acoustic' Data!

## Purpose

ibmAcousticR facilitates making a connection to the IBM Acoustic email campaign management API and executing various queries. The IBM Acoustic API documentation is available at <https://developer.ibm.com/customer-engagement/docs/>.


## Installation

The development version can be installed from GitHub: `devtools::install_github("chrisumphlett/ibmAcousticR")` .

The package is available on CRAN: `install.packages("ibmAcousticR")`.


## Usage

Before utilizing this package you must obtain the proper credentials from IBM Acoustic. This will require someone with administrator privileges to create or obtain the Client Id and Client Secret and to create individual Refresh Tokens.

Once you have these parameters you should first connect to the API using `acoustic_auth()`.

> access_token <- acoustic_auth(org_client_id = "abc", org_client_secret = "xyz", my_refresh_token = "123")

If you are able to authenticate you will obtain an access token. Access tokens are granted temporarily, expiring after four hours. 

The access token is then used to provide the authentication when you submit calls to the API to obtain data. You will also need to know your "pod number" (the number that appears in the URL of your Silverpop portal). 

Some API calls will return data back to your R session; others will submit a job that will allow you to download your file from the Silverpop portal (you must be able to log in). You can get the report within the portal from Resources -> Data Jobs. You may optionally have these sent to an SFTP server provided by Acoustic.

Currently the package has two functions for obtaining data, `get_all_contacts()` and `get_contact_list()`. Each submits a job to be retrieved from the Acoustic/Silverpop portal. See the function documentation for more information. A simple example of using it is below:

> job_id <- get_all_contacts(1, access_token)
> get_job_status(1, access_token, "123456789")

You can get a list of programs which can be joined to the email events using the `program_id` with `get_programs()`:

> get_programs(1, access_token, "2020-01-01", "2020-05-31")

NEWS.md
# ibmAcousticR 0.2.1.9000

* `styler` used to clean up formatting.

# ibmAcousticR 0.2.1
* Added a RETRY() to safely retry an API request certain number of times before returning a error code. 
* Made minor edits to the DESCRIPTION file by adding a contributor.

# ibmAcousticR 0.2.0
* Users may choose between submitting jobs to be manually downloaded from the Acoustic/Silverpop portal or sent to the FTP server.
* Changed end date parameter in `get_contact_list()` to make it inclusive of that date.

# ibmAcousticR 0.1.4
* `get_contact_list()` submits a job to get an export of a contact list or database.
* Utility functions added and/or modified to improve handling of errors with API requests.

# ibmAcousticR 0.1.3

* `get_programs()` returns a data frame with all programs created between specific dates.
* Added `program_id` as column in the get contact events results.

# ibmAcousticR 0.1.2

* Updated URLs as IBM announced it will be deprecating the old URLs on 2020-06-30.
* Minor edits per CRAN reviewer.
* Released to CRAN.

# ibmAcousticR 0.1.1

* Added option to choose between event and sent date for filtering for raw contact events.
* DESCRIPTION file minor edits per CRAN reviewer feedback.
* Example for non-exported function removed per CRAN reviewer feedback.

# ibmAcousticR 0.1.0

* Tested and released to GitHub.
* Initial submission to CRAN.

# ibmAcousticR 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.
* Created first group of functions that allow one to authenticate and query for email contact event data.


R/utils.R
#' Get Job Id of Submitted Job
#'
#' This function is called by other functions that submit jobs to
#' the Acoustic/Silverpop API. It extracts the Job Id from the XML
#' returned by the API call.
#'
#' Job results are available as exports in the Silverpop portal by
#' going to Resources -> Data Jobs.
#'
#' It is not recommended that these authentication parameters be
#' stored directly in your code. There are various methods and
#' packages available that are more secure; this package does not
#' require you to use any one in particular.
#'
#' @param request_obj Name of the object returned from API call,
#' should always be "request".
#' @param path XML path to the job id.
#'
#' @importFrom httr "content"
#' @importFrom XML "xmlParse"
#' @importFrom XML "xmlValue"
#' @importFrom XML "xpathSApply"
#'
#' @return A vector with the Job Id.
#'
#' @keywords internal

get_job_id <- function(request_obj, path) {
  # Extract the XML from the request results
  request_content <- httr::content(request_obj, "text", encoding = "ISO-8859-1")
  request_xml <- XML::xmlParse(request_content)

  # Return the job id
  job_id <- XML::xpathSApply(request_xml, path, XML::xmlValue)
  message(paste0("Submit was successful, Job Id: ", job_id))
  return(job_id)
}

#' Check Request Status
#'
#' This function is called by other functions that submit jobs to
#' the Acoustic/Silverpop API. It checks the status code
#' returned and tells the user if there was an error code and
#' exits the function.
#'
#' @return If status code is not 200, a message to console
#'
#' @keywords internal

check_request_status <- function(request_obj) {
  if (request_obj$status_code == 401) {
    message(paste0(
      "There was a 401 error.",
      "Do you need to refresh your access token?"
    ))
    stop_quietly()
  } else if (request_obj$status_code != 200) {
    message(paste0(
      "There was an authentication error: ",
      request_obj$status_code
    ))
    stop_quietly()
  }
}

#' Stop Function Quietly
#'
#' Quit a function execution without printing error messages. The
#' idea came from a Stack Overflow answer
#' https://stackoverflow.com/questions/14469522/stop-an-r-program-without-error.
#'
#' @return Exits a function.
#'
#' @keywords internal

stop_quietly <- function() {
  opt <- options(show.error.messages = FALSE)
  on.exit(options(opt))
  stop()
}

#' Detect Faulty XML Request
#'
#' Searches the results content for the tag "<FaultString>". If it
#' is found it gives the user a message and exits the function.
#'
#' @importFrom httr "content"
#'
#' @return Message to the console.
#'
#' @keywords internal

check_for_faulty_xml <- function(request_obj) {
  if (length(grep("<FaultString>", httr::content(request_obj, "text",
    encoding = "ISO-8859-1"
  ))) >= 1) {
    message("Faulty XML request. Check your parameters. Full result text:")
    message(httr::content(request_obj, "text", encoding = "ISO-8859-1"))
    stop_quietly()
  }
}


R/get_programs.R
#' Get List of Programs
#'
#' Get list of all programs in a particular date range. Prior
#' to attempting this you must authenticate and obtain an
#' access token.
#'
#' @param pod_number Pod number is the number in the URL, e.g. 1 in
#' engage1.silverpop.com.
#' @param session_access_token Access token obtained during this session.
#' @param start_date Filter for programs created on or after this date.
#' @param end_date Filter for programs created on or before this date.
#'
#' @importFrom jsonlite "fromJSON"
#' @importFrom httr "RETRY"
#' @importFrom httr "POST"
#' @importFrom httr "content"
#' @importFrom httr "add_headers"
#' @importFrom XML "xmlParse"
#' @importFrom XML "getNodeSet"
#' @importFrom XML "xmlToDataFrame"
#'
#' @return A data frame with the programs and program details.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' access_token <- acoustic_auth(
#'   org_client_id = "abc",
#'   org_client_secret = "xyz",
#'   my_refresh_token = "123"
#' )
#'
#' get_programs(1, access_token, "2020-01-01", "2020-05-31")
#' }
#'
get_programs <- function(pod_number, session_access_token, start_date,
                         end_date) {

  # Reformat the dates
  start_date2 <- as.character(format(as.Date(start_date), "%m/%d/%Y"))
  end_date2 <- as.character(format(as.Date(end_date), "%m/%d/%Y"))

  # Build the XML request
  xml_parameters <- paste0(
    "
    <Envelope>
      <Body>
        <GetPrograms>
          <INCLUDE_ACTIVE>True</INCLUDE_ACTIVE>
          <INCLUDE_INACTIVE>True</INCLUDE_INACTIVE>
          <CREATED_DATE_RANGE>",
    "<BEGIN_DATE>", start_date2, "</BEGIN_DATE>",
    "<END_DATE>", end_date2, "</END_DATE>",
    "</CREATED_DATE_RANGE>
        </GetPrograms>
      </Body>
    </Envelope>"
  )

  # Submit the request
  request <- httr::RETRY("POST",
    url = paste0(
      "https://api-campaign-us-", pod_number,
      ".goacoustic.com/XMLAPI"
    ),
    httr::add_headers(
      "Content-Type" =
        "text/xml;charset=utf-8",
      "Authorization" =
        paste0(
          "Bearer ",
          session_access_token
        )
    ),
    body = xml_parameters,
    encode = "json",
    times = 4,
    pause_min = 10,
    terminate_on = NULL,
    terminate_on_success = TRUE,
    pause_cap = 5
  )

  check_request_status(request)
  check_for_faulty_xml(request)

  # Extract and return the job status
  request_content <- httr::content(request, "text", encoding = "ISO-8859-1")
  request_xml <- XML::xmlParse(request_content)
  request_df <- XML::xmlToDataFrame(nodes = XML::getNodeSet(
    request_xml,
    "//PROGRAM"
  ))
  return(request_df)
}


R/get_job_status.R
#' Get the Status of a Submitted Job
#'
#' Prior to attempting this you must authenticate and obtain an
#' access token, and then submit a call that is processed as a
#' job to retrieve from the Acoustic portal. The function used
#' to submit that job will provide the Job Id.
#'
#' @param pod_number Pod number is the number in the URL, e.g.
#' engage1.silverpop.com.
#' @param session_access_token Access token obtained during this session.
#' @param desired_job_id Id for job for which you want the status.
#'
#' @importFrom jsonlite "fromJSON"
#' @importFrom httr "RETRY"
#' @importFrom httr "POST"
#' @importFrom httr "content"
#' @importFrom httr "add_headers"
#' @importFrom XML "xmlParse"
#' @importFrom XML "xmlValue"
#' @importFrom XML "xpathSApply"
#'
#' @return A vector with the session's access token.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' access_token <- acoustic_auth(
#'   org_client_id = "abc",
#'   org_client_secret = "xyz",
#'   my_refresh_token = "123"
#' )
#'
#' job_id <- get_all_contacts(access_token)
#' get_job_status(1, access_token, "123456789")
#' }
#'
get_job_status <- function(pod_number, session_access_token, desired_job_id) {

  # Build the XML request
  xml_parameters <- paste0("
    <Envelope>
      <Body>
      <GetJobStatus>
      <JOB_ID>", desired_job_id, "</JOB_ID>
      </GetJobStatus>
      </Body>
    </Envelope>
    ")

  # Submit the request
  request <- httr::RETRY("POST",
    url = paste0(
      "https://api-campaign-us-", pod_number,
      ".goacoustic.com/XMLAPI"
    ),
    httr::add_headers(
      "Content-Type" =
        "text/xml;charset=utf-8",
      "Authorization" =
        paste0(
          "Bearer ",
          session_access_token
        )
    ),
    body = xml_parameters,
    encode = "json",
    times = 4,
    pause_min = 10,
    terminate_on = NULL,
    terminate_on_success = TRUE,
    pause_cap = 5
  )

  check_request_status(request)
  check_for_faulty_xml(request)

  # Extract and return the job status
  request_content <- httr::content(request, "text", encoding = "ISO-8859-1")
  request_xml <- XML::xmlParse(request_content)
  job_status <- XML::xpathSApply(
    request_xml,
    "//Envelope/Body/RESULT/JOB_STATUS",
    XML::xmlValue
  )
  return(job_status)
}


R/get_all_contacts.R
#' Get Export of All Email Contact Events
#'
#' This function submits a job to Acoustic that exports all email contact
#' events. Various criteria are available to filter the export. Some,
#' but not all, of these have been built into the parameters of this
#' function. Reading the IBM Acoustic documentation is useful:
#' https://developer.ibm.com/customer-engagement/tutorials/
#' export-raw-contact-events/
#'
#' The date type is set to EVENT by default. If you filter by the sent
#' date you may not get all applicable events, as some events (a future
#' click) will not yet have happened. If you do filter by SENT date and
#' are incrementally updating your data you should plan to go back and
#' retroactively update past dates.
#'
#' Job results are available as exports in the Silverpop portal by
#' going to Resources -> Data Jobs.
#'
#' It is not recommended that these authentication parameters be
#' stored directly in your code. There are various methods and
#' packages available that are more secure; this package does not
#' require you to use any one in particular.
#'
#' @param pod_number Pod number is the number in the URL, e.g.
#' engage1.silverpop.com.
#' @param session_access_token Access token obtained during this session.
#' @param start_date Filter for emails sent on or after this date.
#' @param end_date Filter for emails sent on or before this date.
#' @param date_type Select whether the date filters should be on the event
#' date or the email sent date ("EVENT" or "SENT").
#' @param event_types There are 18 different events. By default all event
#' types are returned. This parameter takes XML arguments where you can
#' override the default and specify all of the events you want. See the
#' Acoustic documentation for the full list.
#' @param export_format Acoustic provides three delimiter file types:
#' 0 (CSV), 1 (PIPE), or 2 (TAB). CSV is the default used here.
#' @param move_to_ftp If TRUE (default is FALSE) will send files to SFTP
#' server instead of being able to download manually from the portal.
#' @param exclude_deleted Do you want to exclude contacts that have been
#' deleted, can be TRUE/FALSE. Per Acoustic, "Inclusion of this
#' element can greatly decrease the time to generate the metrics file and
#' is useful whenever metrics for deleted contacts are not required."
#' @param optional_columns Do you want to include six optional columns
#' in the results, can be TRUE/FALSE. These columns are the mailing name,
#' mailing subject, from email address, from email name, CRM campaign Id,
#' and program Id.
#' @param file_name_prefix Optional argument that should be used if you
#' want to add a particular prefix to the file that you will download
#' from your portal.
#' @param confirm_email Optional argument to specify an email address
#' where IBM will let you know when the job has completed.
#'
#' @importFrom jsonlite "fromJSON"
#' @importFrom httr "RETRY"
#' @importFrom httr "POST"
#' @importFrom httr "content"
#' @importFrom httr "add_headers"
#'
#' @return A vector with the Job Id.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' access_token <- acoustic_auth(
#'   org_client_id = "abc",
#'   org_client_secret = "xyz",
#'   my_refresh_token = "123"
#' )
#'
#' job_id <- get_all_contacts(pod_number, access_token,
#'   "2020-01-01", "2020-01-05",
#'   event_types = "<CLICKS/>",
#'   1, exclude_deleted = TRUE, optional_columns = TRUE
#' )
#' }
#'
get_all_contacts <- function(pod_number, session_access_token, start_date,
                             end_date, date_type = "EVENT",
                             event_types = "<ALL_EVENT_TYPES/>",
                             export_format = 0, move_to_ftp = FALSE,
                             exclude_deleted = FALSE, optional_columns = TRUE,
                             file_name_prefix = "", confirm_email = "") {

  # Reformat the dates
  start_date2 <- as.character(format(as.Date(start_date), "%m/%d/%Y"))
  end_date2 <- as.character(format(as.Date(end_date), "%m/%d/%Y"))

  # Build the XML request
  xml_parameters <- paste0(
    "
    <Envelope>
      <Body>
        <RawRecipientDataExport>",
    ifelse(date_type == "EVENT", paste0(
      "<EVENT_DATE_START>", start_date2, "</EVENT_DATE_START>",
      "<EVENT_DATE_END>", end_date2, "</EVENT_DATE_END>"
    ), paste0(
      "<SEND_DATE_START>", start_date2, "</SEND_DATE_START>",
      "<SEND_DATE_END>", end_date2, "</SEND_DATE_END>"
    )),
    "<EXPORT_FORMAT>", export_format, "</EXPORT_FORMAT>",
    ifelse(move_to_ftp == FALSE, "", "<MOVE_TO_FTP/>"),
    ifelse(file_name_prefix != "", paste0(
      "<EXPORT_FILE_NAME>",
      file_name_prefix,
      "</EXPORT_FILE_NAME>"
    ), ""),
    ifelse(confirm_email != "", paste0(
      "<EMAIL>", confirm_email,
      "</EMAIL>"
    ), ""),
    event_types,
    ifelse(exclude_deleted == TRUE, "<EXCLUDE_DELETED/>", ""),
    ifelse(optional_columns == TRUE,
      "<RETURN_MAILING_NAME/>
            <RETURN_SUBJECT/>
            <RETURN_FROM_ADDRESS/>
            <RETURN_FROM_NAME/>
            <RETURN_CRM_CAMPAIGN_ID/>
            <RETURN_PROGRAM_ID/>", ""
    ),
    "</RawRecipientDataExport>
      </Body>
    </Envelope>
    "
  )

  # Submit the request
  request <- httr::RETRY("POST",
    url = paste0(
      "https://api-campaign-us-", pod_number,
      ".goacoustic.com/XMLAPI"
    ),
    httr::add_headers(
      "Content-Type" =
        "text/xml;charset=utf-8",
      "Authorization" =
        paste0(
          "Bearer ",
          session_access_token
        )
    ),
    body = xml_parameters,
    encode = "json",
    times = 4,
    pause_min = 10,
    terminate_on = NULL,
    terminate_on_success = TRUE,
    pause_cap = 5
  )

  check_request_status(request)
  check_for_faulty_xml(request)

  # Get and return the Job Id
  job_id <- get_job_id(request, path = "//Envelope/Body/RESULT/MAILING/JOB_ID")
  return(job_id)
}


R/get_contact_list.R
#' Get Export of a Database or contact List
#'
#' This function submits a job to Acoustic that exports a particular
#' database or contact list based on the list id.  Various criteria
#' are available to filter the export. Some, but not all, of these
#' have been built into the parameters of this function. Reading the
#' IBM Acoustic documentation is useful:
#' https://developer.ibm.com/customer-engagement/tutorials/
#' export-from-a-database/
#'
#' Job results are available as exports in the Silverpop portal by
#' going to Resources -> Data Jobs.
#'
#' It is not recommended that these authentication parameters be
#' stored directly in your code. There are various methods and
#' packages available that are more secure; this package does not
#' require you to use any one in particular.
#'
#' @param pod_number Pod number is the number in the URL, e.g.
#' engage1.silverpop.com.
#' @param session_access_token Access token obtained during this session.
#' @param list_id Acoustic id for the database or contact list (string).
#' @param start_date Filter for emails sent on or after this date.
#' @param end_date Filter for emails sent on or before this date.
#' @param export_format Acoustic provides three delimiter file types:
#' CSV, PIPE, TAB. CSV is the default used here.
#' @param move_to_ftp If TRUE (default is FALSE) will send files to SFTP server
#' instead of being able to download manually from the portal.
#' @param confirm_email Optional argument to specify an email address
#' where IBM will let you know when the job has completed.
#'
#' @importFrom jsonlite "fromJSON"
#' @importFrom httr "RETRY"
#' @importFrom httr "POST"
#' @importFrom httr "content"
#' @importFrom httr "add_headers"
#'
#' @return A vector with the Job Id.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' access_token <- acoustic_auth(
#'   org_client_id = "abc",
#'   org_client_secret = "xyz",
#'   my_refresh_token = "123"
#' )
#'
#' job_id <- get_contact_list(
#'   pod_number, access_token, list_id,
#'   "2020-01-01", "2020-01-05", "PIPE"
#' )
#' }
#'
get_contact_list <- function(pod_number, session_access_token, list_id,
                             start_date, end_date, export_format = "CSV",
                             move_to_ftp = FALSE, confirm_email = "") {

  # Reformat the dates
  start_date2 <- as.character(format(as.Date(start_date), "%m/%d/%Y %H:%M:%S"))
  end_date2 <- as.character(format(as.Date(end_date) + 1, "%m/%d/%Y %H:%M:%S"))

  # Build the XML request
  xml_parameters <- paste0(
    "
    <Envelope>
      <Body>
        <ExportList>",
    "<DATE_START>", start_date2, "</DATE_START>",
    "<DATE_END>", end_date2, "</DATE_END>",
    "<EXPORT_FORMAT>", export_format, "</EXPORT_FORMAT>",
    "<LIST_ID>", list_id, "</LIST_ID>",
    ifelse(move_to_ftp == FALSE, "<ADD_TO_STORED_FILES/>", ""),
    ifelse(confirm_email != "", paste0(
      "<EMAIL>", confirm_email,
      "</EMAIL>"
    ), ""),
    "<INCLUDE_LEAD_SOURCE/>
          <INCLUDE_RECIPIENT_ID/>
          <EXPORT_TYPE>ALL</EXPORT_TYPE>",
    "</ExportList>
      </Body>
    </Envelope>"
  )

  # Submit the request
  request <- httr::RETRY("POST",
    url = paste0(
      "https://api-campaign-us-", pod_number,
      ".goacoustic.com/XMLAPI"
    ),
    httr::add_headers(
      "Content-Type" =
        "text/xml;charset=utf-8",
      "Authorization" =
        paste0(
          "Bearer ",
          session_access_token
        )
    ),
    body = xml_parameters,
    encode = "json",
    times = 4,
    pause_min = 10,
    terminate_on = NULL,
    terminate_on_success = TRUE,
    pause_cap = 5
  )

  check_request_status(request)
  check_for_faulty_xml(request)

  # Get and return the Job Id
  job_id <- get_job_id(request, path = "//Envelope/Body/RESULT/JOB_ID")
  return(job_id)
}


R/acoustic_auth.R
#' Connect to API and Obtain Access Token
#'
#' Prior to attempting this you must have a Client Id, Client
#' Secret and Refresh Token. The first two are assigned on
#' an organization level; the latter must be created by
#' someone with an admin role in Acoustic and assigned to you.
#'
#' Access tokens expire after four hours. Thus, this function should
#' be run each time you utilize the package and may need to be
#' re-called periodically if you have a session open for a long
#' duration.
#'
#' It is not recommended that these authentication parameters be
#' stored directly in your code. There are various methods and
#' packages available that are more secure; this package does not
#' require you to use any one in particular.
#'
#' More information on this available at https://developer.ibm.com/
#' customer-engagement/tutorials/
#' getting-started-oauth-watson-campaign-automation/ .
#'
#' @param org_client_id Organization's Client Id.
#' @param org_client_secret Organization's Client Secret.
#' @param my_refresh_token Your personal Refresh Token.
#' @param pod_number Pod number is the number in the URL, e.g.
#' engage1.silverpop.com.
#'
#' @importFrom jsonlite "fromJSON"
#' @importFrom httr "RETRY"
#' @importFrom httr "POST"
#' @importFrom httr "content"
#' @importFrom httr "add_headers"
#'
#' @return A vector with the session's access token.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' access_token <- acoustic_auth(
#'   org_client_id = "abc",
#'   org_client_secret = "xyz",
#'   my_refresh_token = "123"
#' )
#' }
#'
acoustic_auth <- function(org_client_id, org_client_secret, my_refresh_token,
                          pod_number) {

  # Store the credentials in a list
  body_parameters <- list(
    grant_type = "refresh_token",
    client_id = org_client_id,
    client_secret = org_client_secret,
    refresh_token = my_refresh_token
  )

  # Submit the request
  request <- httr::RETRY("POST",
    url = paste0(
      "https://api-campaign-us-", pod_number,
      ".goacoustic.com/oauth/token"
    ),
    httr::add_headers(
      "Content-Type" =
        "application/x-www-form-urlencoded"
    ),
    body = body_parameters,
    encode = "form",
    times = 4,
    pause_min = 10,
    terminate_on = NULL,
    terminate_on_success = TRUE,
    pause_cap = 5
  )

  check_request_status(request)
  check_for_faulty_xml(request)

  # Get and return the access_token
  request_content <- httr::content(request, "text", encoding = "ISO-8859-1")
  content_json <- jsonlite::fromJSON(request_content, flatten = TRUE)
  access_token <- content_json$access_token
  return(access_token)
}
